### PR TITLE
fix: slow post message

### PR DESCRIPTION
**Pull request by 2xAA**
Limits meyda source input links for performance boost.
Adds new input link type "mutation" to further increase overall performance.

Adds a queue for syncing commits between the worker and renderer threads - sending lots of messages via postMessage is synchronous, it's better to send fewer messages with more content than lots of messages. we're talking around 40+ messages per 16ms here, so combining all these updates and syncing once per frame tick should help.

### CODE REVIEW

File: src/application/worker/index.worker.js [hunk 1]
```diff
@@ -293,11 +363,20 @@ async function start() {
         return JSON.stringify(preset);
       }
 
+      if (identifier === "groups/REPLACE_GROUP_MODULES") {
```
Comment (2xAA) [inline src/application/worker/index.worker.js]: Oops, need to remove this.

File: src/application/worker/index.worker.js [hunk 2]
```diff
@@ -293,11 +363,20 @@ async function start() {
...
+        console.log("message recieved");
+      }
+
       /**
        * @todo Don't JSON parse and stringify
        */
       const value = await store[type](identifier, payload);
 
+      if (identifier === "groups/REPLACE_GROUP_MODULES") {
```
Comment (2xAA) [inline src/application/worker/index.worker.js]: Oops, need to remove this too 😅

**Review by 2xAA — COMMENTED** [review]
Need to clear up some logging!

File: src/application/worker/index.worker.js [hunk 3]
```diff
@@ -9,42 +9,105 @@ async function start() {
...
+      let payloadMatches = false;
+
+      if (match.payload) {
+        const matchPayloadKeys = Object.keys(match.payload);
+        payloadMatches = matchPayloadKeys.every(key => {
+          const value = match.payload[key];
+          return value === mutationPayload[key];
+        });
+      } else {
+        payloadMatches = true;
```
Comment (TimPietrusky) [inline src/application/worker/index.worker.js:83]: Why is this `true` if there is no payload?

File: src/application/worker/index.worker.js [hunk 4]
```diff
@@ -9,42 +9,105 @@ async function start() {
...
+
+      const { type: linkType, location: linkLocation, match } = link;
+
+      if (linkType !== "mutation" || match.type !== mutationType) {
+        continue;
+      }
+
+      let payloadMatches = false;
+
+      if (match.payload) {
```
Comment (TimPietrusky) [inline src/application/worker/index.worker.js:76]: What is `match`?

File: src/application/worker/index.worker.js [hunk 5]
```diff
@@ -9,42 +9,105 @@ async function start() {
...
+        const matchPayloadKeys = Object.keys(match.payload);
+        payloadMatches = matchPayloadKeys.every(key => {
+          const value = match.payload[key];
+          return value === mutationPayload[key];
+        });
+      } else {
+        payloadMatches = true;
+      }
+
+      if (!payloadMatches) {
```
Comment (TimPietrusky) [inline src/application/worker/index.worker.js:86]: If we have no payloadMatch, we continue? Why do we check if the payload is matching anyway? To see if there was a change?

File: src/application/worker/index.worker.js [hunk 6]
```diff
@@ -9,42 +9,105 @@ async function start() {
...
+      (mutationPayload.key !== "props" || mutationPayload.key !== "meta")
     ) {
       return;
     }
 
+    const {
+      inputs: { inputs, inputLinks }
+    } = store.state;
+
+    // Update mutation type Input Links
```
Comment (TimPietrusky) [inline src/application/worker/index.worker.js:56]: What is an input link of type `mutation`?

File: src/application/worker/index.worker.js [hunk 7]
```diff
@@ -9,42 +9,105 @@ async function start() {
...
+    for (let i = 0; i < inputLinksLength; ++i) {
+      const link = mutationTypeInputLinks[i];
+      const inputId = link.id;
+      const bind = inputs[inputId];
+
+      const { type, location, data } = bind;
+
+      const { type: linkType, location: linkLocation, match } = link;
+
+      if (linkType !== "mutation" || match.type !== mutationType) {
```
Comment (TimPietrusky) [inline src/application/worker/index.worker.js]: Why is this needed? As we are already filtering all values of `link.type  === "mutation"` in line 57?

File: src/application/worker/loop.js [hunk 8]
```diff
@@ -69,8 +69,21 @@ function loop(delta, features, fftOutput) {
...
       args: linkArguments,
       min,
-      max
+      max,
+      source
     } = link;
+
+    const moduleId = store.state.inputs.inputs[inputId].data.moduleId;
+
+    if (
```
Comment (TimPietrusky) [inline src/application/worker/loop.js:81]: Please add a comment to make clear that the links of a module are not updated when the module is not enabled?

**Review by TimPietrusky — CHANGES_REQUESTED** [review]
@2xAA thank you very much for this, I'm super happy that we could address this annoying issue. 

I have some questions regarding the documentation and a general question.

I also wonder if we have to update these too regarding the new `mutation` type:

* https://github.com/vcync/modV/blob/322385e6a59aa075c80c240a783478f95835853e/src/components/InputLinkComponents/Expression.vue
* https://github.com/vcync/modV/blob/322385e6a59aa075c80c240a783478f95835853e/src/components/InputLinkComponents/Tween.vue

File: src/application/worker/index.worker.js (hunk 7 above)
Comment (2xAA) [reply src/application/worker/index.worker.js]: Good catch, I added the filter later. I'll remove this

File: src/application/worker/index.worker.js (hunk 3 above)
Comment (2xAA) [reply src/application/worker/index.worker.js:83]: Because it's possible for mutations to have no payload, or we just want to update when a mutation, e.g. `beats/SET_KICK`, is committed.

Example here: https://github.com/vcync/modV/blob/322385e6a59aa075c80c240a783478f95835853e/src/components/InputLinkComponents/AudioFeatures.vue#L123-L132

File: src/application/worker/index.worker.js (hunk 6 above)
Comment (2xAA) [reply src/application/worker/index.worker.js:56]: There are multiple types of input links.
* all input links
  * Params:
    * inputId
      * String. The ID of the input to link to
    * ?source
      * String. Identifies what created and/or manages this input link
    * ?min
      * Number. Minimum expected value, will scale the incoming value to the input's min/max if min and max are set
    * ?max
      * Number. Maximum expected value, will scale the incoming value to the input's min/max if min and max are set
* getter
  *  _*:warning: updates every frame*_
  * Reads a getter in the store, possibly with arguments, defined in `?args`, and applies that to the input link defined with `inputId`
  * Params:
    * location
      * String. location of the getter in the store, e.g. `meyda/getFeature`
    * ?args
      * Array. arguments for a getter that is a function, e.g. `["energy"]`
* mutation
  * Waits for a mutation type, defined with `match.type`, possibly checks the mutation payload for specific unique parameters, defined on `match.?payload`, and applies a store value, specified with `location`, to the input link defined with `inputId`
  * Params:
    * location
      * String. location of the value to read in the store, e.g. `midi.devices[1].channelData[1][144]`
    * match
      * Object. Parameters of the mutation to match. Has keys:
        * type
          * String. Mutation type from the store, e.g. `midi/WRITE_DATA`
        * ?payload
          * Object. Key-value pairs to match in the mutation payload. e.g.
          ```JavaScript
          {
            id: `${id}-${name}-${manufacturer}`,
            channel,
            type
          } 
          ```
* state
  *  _*:warning: updates every frame*_
  * Reads a location in the store and applies that to the input link defined with `inputId`
  * Params:
    * location
      * String. location of the value to read in the store, e.g. `midi.devices[1].channelData[1][144]`

state and getter links are updated in loop.js
https://github.com/vcync/modV/blob/322385e6a59aa075c80c240a783478f95835853e/src/application/worker/loop.js#L89

File: src/application/worker/index.worker.js (hunk 4 above)
Comment (2xAA) [reply src/application/worker/index.worker.js:76]: Explained here: https://github.com/vcync/modV/pull/623/files/322385e6a59aa075c80c240a783478f95835853e#r692313186

File: src/application/worker/index.worker.js (hunk 5 above)
Comment (2xAA) [reply src/application/worker/index.worker.js:86]: The payload may need to match to check for things like IDs or other unique parameters so we know the correct inputLink will be updated.

If we didn't check for the right ID on the MIDI input links, all MIDI input links would update on any MIDI input which would be wasteful in terms of postMessage.

**Comment by 2xAA** [comment]
> @2xAA thank you very much for this, I'm super happy that we could address this annoying issue.
> 
> I have some questions regarding the documentation and a general question.
> 
> I also wonder if we have to update these too regarding the new `mutation` type:
> 
> * https://github.com/vcync/modV/blob/322385e6a59aa075c80c240a783478f95835853e/src/components/InputLinkComponents/Expression.vue
> * https://github.com/vcync/modV/blob/322385e6a59aa075c80c240a783478f95835853e/src/components/InputLinkComponents/Tween.vue

@TimPietrusky Thanks for the review.

Expressions are handled independently of Input Links, they are already reactive.

Tweens could be updated, but they update every frame anyway so I don't think there would be any performance gain.
We may want to end their updates when the module is not enabled though, like we did with meyda source Input Links.

**Comment by TimPietrusky** [comment]
@2xAA I was just trying to convert my old preset to use `mutation` and I'm wondering if we actually need the `inputLinks.*.location` anymore? As the same data is also in `match` and I also couldn't find any code that is using the `location`. Should this be removed from the saved preset?

**Comment by 2xAA** [comment]
Location is still required at the moment. Please add this.

File: src/application/worker/index.worker.js (hunk 6 above)
Comment (TimPietrusky) [reply src/application/worker/index.worker.js:56]: @2xAA thanks for this explanation!

**Comment by 2xAA** [comment]
> @2xAA I was just trying to convert my old preset to use `mutation` and I'm wondering if we actually need the `inputLinks.*.location` anymore? As the same data is also in `match` and I also couldn't find any code that is using the `location`. Should this be removed from the saved preset?

@TimPietrusky, `inputLinks.*.location` is required and for API consistency we should keep this.

To explain my choice on keeping location, the only other way we could implement this, that I can think of, is to specify the key within the payload to get the value from and write that to the Input.
Right now, not every mutation has an Object as the payload, so it may serve us well to just keep the same structure.
This would also limit the possibilities of Input Links in the future as you may want to read data from another location based on a totally different mutation type.

If you agree with this then I think I will write up the explainer in one of the previous comments as JSDoc and we can go ahead and merge?

**Comment by TimPietrusky** [comment]
@2xAA yes, that would be super perfect! Thanks ❤️

**Review by TimPietrusky — APPROVED** [review]
Thank you very much <3
